### ISSUE TITLE
[STORM-67] Provide API for spouts to know how many pending messages there are

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#3909]
<p><a href="https://github.com/nathanmarz/storm/issues/343" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/issues/343</a></p>

<p>This would be useful in case you want to take special action in the spout like drop messages</p>

<p>-----------------<br/>
Discmt: Hi, I'd like to try and take a crack at this if it's still relevant. I'm not exactly sure what it's asking for though. It seems to me an implementation for knowing how many pending messages there are for a spout depends on where the spout is getting it's information from, which makes me sure I'm missing something.</p>

<p>-----------------<br/>
revans2: The spout code in backtype/storm/daemon/executor.clj is already keeping track of the pending tuples if acking is enabled. If acking is disabled nothing is pending.</p>

<p>defmethod mk-threads :spout <span class="error">&#91;executor-data task-datas&#93;</span></p>

<p>defines pending as a RotatingMap which maps all of the storm internal tuple ids to the message id objects passed in by the spout when it first emitted the tuple. The hardest part should be getting pending to a place where the ISpoutOutputCollector implementation or where ever the API is, can get access to it.</p>

<p>-----------------<br/>
ptgoetz: @Discmt Yes, this is still relevant and would be nice to have.</p>

<p>The Storm framework asks spouts for tuples by calling the nextTuple() method and keeps track of the tuple tree internally. The underlying data source does not come into play.</p>

<p>As implied by @revans2, one approach would be to add a method to ISpoutOutputCollector such as getPendingCount() that would allow spout implementations to query for the pending count (probably returning -1 if acking is disabled). The tricky part will likely be bridging the gap between executor.clj and the ISpoutOutputCollector implementation(s). I haven't dug very deeply into the code, so off-hand I don't know how hard that would be. A quick search of the code for TOPOLOGY_MAX_PENDING should point you to some of the touch points.</p>

<p>Also keep in mind the dual meaning of TOPOLOGY_MAX_PENDING. In a standard storm topology it represents the maximum number of outstanding tuples. In a trident topology it represents the maximum number of outstanding batches.</p>

<p>-----------------<br/>
Discmt: Hey guys. I've been taking time to look into it, and I feel like I might have an understanding of what exactly it is I need to do. If what @revans2 said is true, and all pending messages are kept within that RotatingMap then this should be somewhat straightforward. I am trying to compile my own storm.jar file right now but I haven't figured how. I tried using build_release.sh in the bin file, but I had no luck. I also tried using lein jar</p>

<p>-----------------<br/>
xumingming: try the following:</p>

<p>lein sub install<br/>
lein install</p>

<p>after these commands are executed, there should be a jar file named storm-xxx.jar in $STORM_HOME/target/.</p>

<p>-----------------<br/>
Discmt: @xumingming . Thanks for the advice. I found that I had Leiningen 1, but the minimum for is Leiningen 2.</p>

<p>-----------------<br/>
xumingming: yeah, storm requires lein 2 to build: <a href="https://github.com/nathanmarz/storm/blob/master/project.clj#L14" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/blob/master/project.clj#L14</a></p>

<p>-----------------<br/>
Discmt: Hi guys. I got my development environment squared away and I can properly build releases now. I use the build_release.sh script. I tried making a change the way @ptgoetz and @revans2 had suggested by adding a method to the output collector to return the pending count. I have some questions about it.</p>

<p>I noticed most of the collector implementations rely on a delegate, or mediator, which I'm assuming is defined here: <a href="https://github.com/nathanmarz/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L504-515" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L504-515</a>. So if I make a add a method to get the size of pending, defined here <a href="https://github.com/nathanmarz/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L408-414" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L408-414</a>, like so:</p>

<p> (SpoutOutputCollector.<br/>
  (reify ISpoutOutputCollector<br/>
    (^int getPendingCount<span class="error">&#91;this&#93;</span> <br/>
      (.size pending)<br/>
      )<br/>
    (^List emit <span class="error">&#91;this ^String stream-id ^List tuple ^Object message-id&#93;</span><br/>
      (send-spout-msg stream-id tuple message-id nil)<br/>
      )<br/>
    (^void emitDirect [this ^int out-task-id ^String stream-id<br/>
       ^List tuple ^Object message-id]<br/>
      (send-spout-msg stream-id tuple message-id out-task-id)<br/>
      )<br/>
    (reportError <span class="error">&#91;this error&#93;</span><br/>
      (report-error error)<br/>
      )))))</p>

<p>I should be good right? Aside from the collectors of the trident spouts which may take more research.</p>

<p>Just so I'm clear, messages are considered pending if they have left the spout and are waiting to be "fully processed", as defined here: <a href="https://github.com/nathanmarz/storm/wiki/Guaranteeing-message-processing" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/wiki/Guaranteeing-message-processing</a> ?</p>

<p>My last question is: does anyone have any suggestions on what would be a good way to test this? i.e. "What kind of topology/scenario should I run".</p>

<p>-----------------<br/>
revans2: The code you pasted above is creating a new SoputOputCollector instance wrapping an ISpoutOutputCollector instance. In order for any java spout to actually get access to the getPendingCount method, you will need to modify ISpoutOutputCollector to have that method defined in it, and also update SpoutOutputCollector, etc to also implement that method delegating to the ISpoutOutputCollector instance.</p>

<p>You are correct about the definition of pending.</p>

<p>Testing is a bit more difficult. You could have a spout that sends out messages to a bolt that does not process them. You could then verify that each time you send out a message the pending count goes up by 1. But to fully test it would would need to have some coordination between the spout and the bolt. This is not impossible but you may need to use global values to do so.</p>

<p>-----------------<br/>
Discmt: @revans2 I got the interface implementation covered. Thank you for the testing suggestion, and hints on how to perform it. It's incredibly helpful. I think that's a good idea, and I'm going to try and do that to test it out.</p>

<p>-----------------<br/>
nathanmarz: There's testing infrastructure already built that can do this kind of tracking. The name in the testing code is "tracked topologies". I'm traveling right now so can't give a link but you should be able to find it. It's used quite a bit in the tests of the acking system.</p>

<p>-----------------<br/>
Discmt: @nathanmarz This is also good news. I'll take a look and use what I can there to speed up the process for me.</p>

<p>-----------------<br/>
Discmt: Hi, I am asking if someone can explain to me how these tracked topologies work.</p>

<p>Particularly I'm confused about these lines here:<br/>
<a href="https://github.com/nathanmarz/storm/blob/master/storm-core/test/clj/backtype/storm/integration_test.clj#L224-245" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/blob/master/storm-core/test/clj/backtype/storm/integration_test.clj#L224-245</a></p>

<p>Looking in integration_test.clj under the function test-acking on line 219 one of the spouts is told to feed a tuple(.feed feeder1 <span class="error">&#91;1&#93;</span>). Then a tracked-wait is called on the topology being tracked to wait for one tuple to be emitted from the spout. Afterwards the checker checks whether or not 0 tuples have been acknowledged, and this what I expect is that one 1 tuple would have been acked because it was emitted from the spout. However, this is not the case. Then later on there checker1 is called again expecting there to be one tuple, but another one was not fed. Furthermore, feeder2, a second spout, was told to feed a tuple, and as expected one was already there when checker2 checked for one ack, but this behavior is not the same as the first checker.</p>

<p>-----------------<br/>
nathanmarz: tracked-wait waits until the entire tree of processing of the tuple has finished. The "checker1" function checks how many tuples on the spout have been acked since last time it was called. The topology is set up so that acknowledgement of tuple trees is delayed until bolts have received multiple tuples. That's why nothing is acked on the spout after the first tuple is emitted, but it is when the second tuple is emitted. If you look at the logic of branching-bolt and agg-bolt you'll see why this is the case.</p>

---
<details><summary><i>Originally reported by <a title="James Xu" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=xumingming">xumingming</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-67" target="_blank">Provide API for spouts to know how many pending messages there are</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Shyam Rajendran" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-07-18T03:56:53+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>I started with the implementation and test assuming the 'pending' RotatingMap defined inside mk-threads [ <a href="https://github.com/apache/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L456" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L456</a> ] gives the right pending ack count.  <br/>
Function 'getPendingCount' was added to ISpoutOutputCollector and SpoutOutputCollector. The required changes for other interfaces and classes were done.  But I am having difficulty in understanding why the pending count always returns 0 [ tested with acking enabled and disabled ]  i.e size of the RotatingMap returns 0.  </p>

<p> (^long getPendingCount<span class="error">&#91;this&#93;</span><br/>
      (.size pending)<br/>
      ) <br/>
was added below <a href="https://github.com/apache/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L560" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/storm-core/src/clj/backtype/storm/daemon/executor.clj#L560</a></p>

<p>The testing was done through simple sysout inside nextTuple() [ example inside TestWordSpout.java) with acking enabled. I was looking to see some pattern initially but found zeros. I then disabled acking but found similar observation. </p>

<p>Would be helpful if somebody can explain the use of pending elsewhere if so. </p>

<p>Changes @ <a href="https://gist.github.com/bourneagain/eab47890aba02e67b1b6" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/bourneagain/eab47890aba02e67b1b6</a></p>

<p>My understanding of how RotatingMap is populated is from <a href="https://github.com/apache/storm/blob/master/docs/documentation/Acking-framework-implementation.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/docs/documentation/Acking-framework-implementation.md</a> and my expectation was to see the size of pending RotatingMap datastructure to keep increasing if acking is diabled. Please correct me if I am wrong. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>My bad, I had not anchored the tuples. <a href="https://storm.apache.org/documentation/Guaranteeing-message-processing.html" class="external-link" target="_blank" rel="nofollow noopener">https://storm.apache.org/documentation/Guaranteeing-message-processing.html</a></p>

<p>I could now see non zero pending RotatingMap size.<br/>
I shall re review the code and issue a pull request after some more testing. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>The getPendingCount presently returns 0 if the acking is disabled. It should probably return -1 as suggested in earlier posts. Working on this change now.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>I see one way to check if acking is enabled or not is to check if the 'send-spout-msg' [ executor.clj ] has been called with a non nil 'message-id' as it depends on the value of  rooted? (and message-id has-ackers?) whose value is not available outside the send-spout-msg function scope. So for now I assume the pending count will not be checked if the user does not enable acking.  </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user bourneagain opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/593" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-67" title="Provide API for spouts to know how many pending messages there are" class="issue-link" data-issue-key="STORM-67"><del>STORM-67</del></a> Provide API for spouts to know how many pending messages there are</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/bourneagain/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/bourneagain/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-67" title="Provide API for spouts to know how many pending messages there are" class="issue-link" data-issue-key="STORM-67"><del>STORM-67</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/593.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #593</p>

<hr />
<p>commit e035ab6f9d20c1ea632c5e999dcf10fc0ce1e2e5<br/>
Author: Shyam Rajendran <srajendran@yahoo-inc.com><br/>
Date:   2015-06-11T17:38:49Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-67" title="Provide API for spouts to know how many pending messages there are" class="issue-link" data-issue-key="STORM-67"><del>STORM-67</del></a> Provide API for spouts to know how many pending messages there<br/>
    are</p>
<ul class="alternate" type="square">
	<li>changes to classes implementing ISpoutOutputCollector</li>
	<li>executor.clj getPendingCount  definition</li>
</ul>


<p>commit 4e832436134ada6639f7520c0445ea05e8130f6c<br/>
Author: Shyam Rajendran <srajendran@yahoo-inc.com><br/>
Date:   2015-06-12T18:04:22Z</p>

<p>    Merge branch 'master' into <a href="https://issues.apache.org/jira/browse/STORM-67" title="Provide API for spouts to know how many pending messages there are" class="issue-link" data-issue-key="STORM-67"><del>STORM-67</del></a></p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user knusbaum commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/593#discussion_r33310534" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593#discussion_r33310534</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/storm/trident/spout/RichSpoutBatchTriggerer.java &#8212;<br/>
    @@ -27,17 +27,12 @@<br/>
     import backtype.storm.tuple.Fields;<br/>
     import backtype.storm.tuple.Values;<br/>
     import backtype.storm.utils.Utils;<br/>
    -import java.util.ArrayList;<br/>
    -import java.util.HashMap;<br/>
    -import java.util.HashSet;<br/>
    -import java.util.List;<br/>
    -import java.util.Map;<br/>
    -import java.util.Random;<br/>
    -import java.util.Set;<br/>
     import storm.trident.topology.TridentBoltExecutor;<br/>
     import storm.trident.tuple.ConsList;<br/>
     import storm.trident.util.TridentUtils;</p>

<p>    +import java.util.*;<br/>
    &#8212; End diff &#8211;</p>

<p>    Let's not import java.util.*.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/593#issuecomment-118856776" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593#issuecomment-118856776</a></p>

<p>    I am working on few more review comments presently and shall update as soon as its available for further review. Thanks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>The pending count is available by exposing the size of the rotating map.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>Waiting on +1s</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/593#issuecomment-120600320" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593#issuecomment-120600320</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/593#issuecomment-121281726" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593#issuecomment-121281726</a></p>

<p>    Thanks @HeartSaVioR .We can have this merged to master whenever we feel appropriate. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/593" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/593</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran" class="user-hover" rel="shyamrajendran">Shyam Rajendran</a> for great work.<br/>
I merged this to master.</p>
